From 320c0acf89acb5950a10816ca7dd083842ff739c Mon Sep 17 00:00:00 2001
From: Thomas Neidhart <tn@apache.org>
Date: Tue, 6 May 2014 21:38:05 +0000
Subject: [PATCH] [COLLECTIONS-521] Fix MultiKeyMap when using two key
 arguments and the second is null. Thanks to Maxime Nay.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/collections/trunk@1592893 13f79535-47bb-0310-9956-ffa450edef68
---
 src/changes/changes.xml                          |  4 ++++
 .../commons/collections4/map/MultiKeyMap.java    |  2 +-
 .../collections4/map/MultiKeyMapTest.java        | 16 ++++++++++++++++
 3 files changed, 21 insertions(+), 1 deletion(-)

diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index 40f22aaf23..82942c3c17 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -22,6 +22,10 @@
   <body>
 
   <release version="4.1" date="TBD" description="">
+    <action issue="COLLECTIONS-521" dev="tn" type="fix" due-to="Maxime Nay">
+      "MultiKeyMap" was throwing a "NullPointerException" for various operations
+      if two key arguments have been used and the second was "null".
+    </action>
     <action issue="COLLECTIONS-522" dev="tn" type="fix" due-to="Erik">
       Updated code example for "PredicatedList".
     </action>
diff --git a/src/main/java/org/apache/commons/collections4/map/MultiKeyMap.java b/src/main/java/org/apache/commons/collections4/map/MultiKeyMap.java
index ee8d4f1272..4c9a993e2f 100644
--- a/src/main/java/org/apache/commons/collections4/map/MultiKeyMap.java
+++ b/src/main/java/org/apache/commons/collections4/map/MultiKeyMap.java
@@ -249,7 +249,7 @@ protected boolean isEqualKey(final AbstractHashedMap.HashEntry<MultiKey<? extend
         return
             multi.size() == 2 &&
             (key1 == multi.getKey(0) || key1 != null && key1.equals(multi.getKey(0))) &&
-            (key2 == multi.getKey(1) || key1 != null && key2.equals(multi.getKey(1)));
+            (key2 == multi.getKey(1) || key2 != null && key2.equals(multi.getKey(1)));
     }
 
     //-----------------------------------------------------------------------
diff --git a/src/test/java/org/apache/commons/collections4/map/MultiKeyMapTest.java b/src/test/java/org/apache/commons/collections4/map/MultiKeyMapTest.java
index ab98d961a0..f9c3c6e9d1 100644
--- a/src/test/java/org/apache/commons/collections4/map/MultiKeyMapTest.java
+++ b/src/test/java/org/apache/commons/collections4/map/MultiKeyMapTest.java
@@ -300,6 +300,22 @@ public void testMultiKeyPut() {
         }
     }
 
+    public void testMultiKeyPutWithNullKey() {
+        final MultiKeyMap<String, String> map = new MultiKeyMap<String, String>();
+        map.put("a", null, "value1");
+        map.put("b", null, "value2");
+        map.put("c", null, "value3");
+        map.put("a", "z",  "value4");
+        map.put("a", null, "value5");
+        map.put(null, "a", "value6");
+        map.put(null, null, "value7");
+        
+        assertEquals(6, map.size());
+        assertEquals("value5", map.get("a", null));
+        assertEquals("value4", map.get("a", "z"));
+        assertEquals("value6", map.get(null, "a"));
+    }
+
     public void testMultiKeyRemove() {
         final MultiKey<K>[] keys = getMultiKeyKeys();
         final V[] values = getSampleValues();
